weasyl/searchtag.py

```diff
@@ -217,3 +260,175 @@ def tag_history(submitid):
+        userid: The userid of the user requesting the list of tags.
+        global_blacklist: Boolean True if requesting tags from the global blacklist. Defaults to False.
+
+    Returns:
+        A list of searchtags (and users for global blacklist requests).
+    """
+    # Only directors can edit the global blacklist; sanity check against the @director_only decorator
+    if global_blacklist and userid not in staff.DIRECTORS:
+        raise WeasylError("InsufficientPermissions")
+
+    # Tags on the global blacklist are being requested
+    if global_blacklist:
+        query = d.engine.execute("""
+            SELECT st.title, lo.login_name
+            FROM searchmapglobalblacklist
+            INNER JOIN searchtag AS st USING (tagid)
+            INNER JOIN login AS lo USING (userid)
+            ORDER BY st.title ASC
+            """).fetchall()
+        return query
+    # User blacklist tags are being requested
+    else:
+        query = d.engine.execute("""
+            SELECT st.title
+            FROM searchmapuserblacklist
+            INNER JOIN searchtag AS st USING (tagid)
+            WHERE userid = %(userid)s
+            ORDER BY st.title ASC
+            """, userid=userid).fetchall()
+        tags = list(x.title for x in query)
+        return tags
+
+
+def query_blacklisted_tags(newtagids, ownerid):
+    """
+    Checks both the user and global searchtag blacklists against added tags, and
+        returns the IDs of any tags matching the STBL, either based on strict
+        match, or regexp based match.
+
+    Parameters:
+        newtagids: The list of added tag ids.
+        ownerid: The userid of the submitted item being checked.
+
+    Returns:
+        blacklisted_tags: The tagids which are blacklisted as a set()
+    """
+    regex_pattern = re.compile(r"""(\w+\*\w+|   # Matches a*a
+                                    \w{2,}\*|   # Matches aa*
+                                    \*\w{2,})   # Matches *aa
+                               """, flags=re.VERBOSE)
+    blacklist_query = d.engine.execute("""
+        SELECT st.tagid, st.title
+        FROM searchmapuserblacklist
+        INNER JOIN searchtag AS st USING (tagid)
+        WHERE userid = %(ownerid)s
+        UNION
+        SELECT st.tagid, st.title
+        FROM searchmapglobalblacklist
+        INNER JOIN searchtag AS st USING (tagid)
+        """, ownerid=ownerid).fetchall()
+    tag_titles = d.engine.execute("""
+        SELECT title, tagid
+        FROM searchtag
+        WHERE tagid = ANY (%(tagids)s)
+        """, tagids=list(newtagids)).fetchall()
+    blacklisted_tags = set()
+
+    for x in blacklist_query:
+        # Determine if the candidate ID is directly present in the newly added IDs.
+        if x.tagid in newtagids:
+            blacklisted_tags.add(x.tagid)
+        # Otherwise we need to parse for wildcards
+        else:
+            # Does this candidate title even have a wildcard?
+            if re.match(regex_pattern, x.title):
```

Comment:
For ``tag.tagid``, at least, [L214/L216](https://github.com/Weasyl/weasyl/pull/179/files#diff-843b15b74ff434962a468d5a89b4ccf4R214) uses the ``tagid``s gathered in this function to strip out any blacklisted tags, based on their IDs from the ``searchmap`` table. So the logic here is:

- Get the user and global blacklists
- Quickly check to see if there is a 1:1 match between ``tagid`` in the blacklists table to the ``tagid`` we are examining.
- Otherwise, check if the tag is a regular expression; if so, test ``i.title`` against the regexp made after converting the STBL entry to a regex pattern.

I can clean up the code a bit here, though; I'll clean it up in the patch.